<h1 style="font-weight: bold;">Titanic Survival With R</h1>


## 1) **Import libraries**

load the libraries to be used in the notebook


In [31]:
library(magrittr)
library(tidyverse)
library(tidymodels)
library(psych)
library(R6)

## 2) **Setting notebook**


In [3]:
options(indent = 4)


## 3) **Load the data**

<p>Load the csv file to pandas dataframe</p>


In [4]:
folder_path <- NULL

if (dir.exists("data/")) {
  folder_path <- "data"
} else {
  stop("Unable find the data")
}


train_path <- file.path(folder_path, "train.csv")
test_path <- file.path(folder_path, "test.csv")


In [5]:
# Set the PassengerId column as index column.
index_col <- "PassengerId"

# Load the train and test data frame.
train_df <- read_csv(train_path)
test_df <- read_csv(test_path)


Rows: 891 Columns: 12
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (5): Name, Sex, Ticket, Cabin, Embarked
dbl (7): PassengerId, Survived, Pclass, Age, SibSp, Parch, Fare

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 418 Columns: 11
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (5): Name, Sex, Ticket, Cabin, Embarked
dbl (6): PassengerId, Pclass, Age, SibSp, Parch, Fare

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


In [6]:
index_col <- "PassengerId"

# Remove the PassengerId column for data index
train_df <- train_df %>% select(-any_of(index_col))
test_df <- test_df %>% select(-any_of(index_col))


## 4) **Data statistics**

<p>Analyse train and test dataframe</p>


### i) Rows in train data

<p>View train dataframe top and bottom rows</p>


In [7]:
# First five rows in train data.
train_df %>% head()


Survived Pclass Name                                                Sex   
1 0        3      Braund, Mr. Owen Harris                             male  
2 1        1      Cumings, Mrs. John Bradley (Florence Briggs Thayer) female
3 1        3      Heikkinen, Miss. Laina                              female
4 1        1      Futrelle, Mrs. Jacques Heath (Lily May Peel)        female
5 0        3      Allen, Mr. William Henry                            male  
6 0        3      Moran, Mr. James                                    male  
  Age SibSp Parch Ticket           Fare    Cabin Embarked
1 22  1     0     A/5 21171         7.2500 NA    S       
2 38  1     0     PC 17599         71.2833 C85   C       
3 26  0     0     STON/O2. 3101282  7.9250 NA    S       
4 35  1     0     113803           53.1000 C123  S       
5 35  0     0     373450            8.0500 NA    S       
6 NA  0     0     330877            8.4583 NA    Q

In [8]:
# First five rows in test data.
train_df %>% tail()


Survived Pclass Name                                     Sex    Age SibSp
1 0        3      Rice, Mrs. William (Margaret Norton)     female 39  0    
2 0        2      Montvila, Rev. Juozas                    male   27  0    
3 1        1      Graham, Miss. Margaret Edith             female 19  0    
4 0        3      Johnston, Miss. Catherine Helen "Carrie" female NA  1    
5 1        1      Behr, Mr. Karl Howell                    male   26  0    
6 0        3      Dooley, Mr. Patrick                      male   32  0    
  Parch Ticket     Fare   Cabin Embarked
1 5     382652     29.125 NA    Q       
2 0     211536     13.000 NA    S       
3 0     112053     30.000 B42   S       
4 2     W./C. 6607 23.450 NA    S       
5 0     111369     30.000 C148  C       
6 0     370376      7.750 NA    Q

### ii) Rows in test data

<p>View test dataframe top and bottom rows</p>


In [9]:
# First five rows in test data.
test_df %>% head()


Pclass Name                                         Sex    Age  SibSp Parch
1 3      Kelly, Mr. James                             male   34.5 0     0    
2 3      Wilkes, Mrs. James (Ellen Needs)             female 47.0 1     0    
3 2      Myles, Mr. Thomas Francis                    male   62.0 0     0    
4 3      Wirz, Mr. Albert                             male   27.0 0     0    
5 3      Hirvonen, Mrs. Alexander (Helga E Lindqvist) female 22.0 1     1    
6 3      Svensson, Mr. Johan Cervin                   male   14.0 0     0    
  Ticket  Fare    Cabin Embarked
1 330911   7.8292 NA    Q       
2 363272   7.0000 NA    S       
3 240276   9.6875 NA    Q       
4 315154   8.6625 NA    S       
5 3101298 12.2875 NA    S       
6 7538     9.2250 NA    S

In [10]:
# Last five rows in test data
test_df %>% tail()


Pclass Name                           Sex    Age  SibSp Parch
1 3      Henriksson, Miss. Jenny Lovisa female 28.0 0     0    
2 3      Spector, Mr. Woolf             male     NA 0     0    
3 1      Oliva y Ocana, Dona. Fermina   female 39.0 0     0    
4 3      Saether, Mr. Simon Sivertsen   male   38.5 0     0    
5 3      Ware, Mr. Frederick            male     NA 0     0    
6 3      Peter, Master. Michael J       male     NA 1     1    
  Ticket             Fare     Cabin Embarked
1 347086               7.7750 NA    S       
2 A.5. 3236            8.0500 NA    S       
3 PC 17758           108.9000 C105  C       
4 SOTON/O.Q. 3101262   7.2500 NA    S       
5 359309               8.0500 NA    S       
6 2668                22.3583 NA    C

### iii) Data description

<p>Get the know the datasets</p>


#### a) Data shapes


In [11]:
# Train data frame shape.
train_shape <- train_df %>% dim()
display_shape <- sprintf(
  "Train shape:\n\trows: %s\n\tcolumns: %s",
  train_shape[1],
  train_shape[2]
)
cat(display_shape)


Train shape:
	rows: 891
	columns: 11

In [12]:
# Test data frame shape.
test_shape <- test_df %>% dim()
display_shape <- sprintf(
  "Test shape:\n\trows: %s\n\tcolumns: %s",
  test_shape[1],
  test_shape[2]
)
cat(display_shape)


Test shape:
	rows: 418
	columns: 10

#### b) Data column names


In [13]:
# Formats the vectors
format_vec <- function(vec) {
  cat("[\n")
  for (i in vec) {
    cat("  ", i, ",\n", sep = "")
  }
  cat("]")
}


In [14]:
columns <- train_df %>% colnames()
format_vec(columns)


[
  Survived,
  Pclass,
  Name,
  Sex,
  Age,
  SibSp,
  Parch,
  Ticket,
  Fare,
  Cabin,
  Embarked,
]

#### c) Data information


In [15]:
# Train data information.
train_df %>% str()


tibble [891 × 11] (S3: tbl_df/tbl/data.frame)
 $ Survived: num [1:891] 0 1 1 1 0 0 0 0 1 1 ...
 $ Pclass  : num [1:891] 3 1 3 1 3 3 1 3 3 2 ...
 $ Name    : chr [1:891] "Braund, Mr. Owen Harris" "Cumings, Mrs. John Bradley (Florence Briggs Thayer)" "Heikkinen, Miss. Laina" "Futrelle, Mrs. Jacques Heath (Lily May Peel)" ...
 $ Sex     : chr [1:891] "male" "female" "female" "female" ...
 $ Age     : num [1:891] 22 38 26 35 35 NA 54 2 27 14 ...
 $ SibSp   : num [1:891] 1 1 0 1 0 0 0 3 0 1 ...
 $ Parch   : num [1:891] 0 0 0 0 0 0 0 1 2 0 ...
 $ Ticket  : chr [1:891] "A/5 21171" "PC 17599" "STON/O2. 3101282" "113803" ...
 $ Fare    : num [1:891] 7.25 71.28 7.92 53.1 8.05 ...
 $ Cabin   : chr [1:891] NA "C85" NA "C123" ...
 $ Embarked: chr [1:891] "S" "C" "S" "S" ...


In [16]:
# Test data information
test_df %>% str()


tibble [418 × 10] (S3: tbl_df/tbl/data.frame)
 $ Pclass  : num [1:418] 3 3 2 3 3 3 3 2 3 3 ...
 $ Name    : chr [1:418] "Kelly, Mr. James" "Wilkes, Mrs. James (Ellen Needs)" "Myles, Mr. Thomas Francis" "Wirz, Mr. Albert" ...
 $ Sex     : chr [1:418] "male" "female" "male" "male" ...
 $ Age     : num [1:418] 34.5 47 62 27 22 14 30 26 18 21 ...
 $ SibSp   : num [1:418] 0 1 0 0 1 0 0 1 0 2 ...
 $ Parch   : num [1:418] 0 0 0 0 1 0 0 1 0 0 ...
 $ Ticket  : chr [1:418] "330911" "363272" "240276" "315154" ...
 $ Fare    : num [1:418] 7.83 7 9.69 8.66 12.29 ...
 $ Cabin   : chr [1:418] NA NA NA NA ...
 $ Embarked: chr [1:418] "Q" "S" "Q" "S" ...


#### d) Describing the data


In [17]:
n_unique <- function(x) length(unique(x))
is_numeric_vec <- function(x, threshold_num = 10) n_unique(x) > threshold_num

desc <- function(data, display_stat = "numeric", threshold_num = 10) {
  if (display_stat == "numeric") {
    numeric_result <- data %>%
      select(where(is.numeric)) %>%
      select(where(is_numeric_vec)) %>%
      describe() %>%
      t()

    return(numeric_result)
  } else if (display_stat == "character") {
    cat_result <- data %>%
      select(where(
        function(vec) {
          is.character(vec) || is.factor(vec) || !is_numeric_vec(
            vec,
            threshold_num = threshold_num
          )
        }
      )) %>%
      lapply(function(vec) {
        c((!(vec %>% is.na())) %>% sum(), n_unique(vec))
      }) %>%
      data.frame()
    row.names(cat_result) <- c("n", "n_unique")

    return(cat_result %>% t())
  }
}


<p style="font-weight: bold; font-style: italic;">Description of numerical columns in train data</p>


In [18]:
train_df %>% desc()


Age         Fare      
vars       1.0000000   2.000000
n        714.0000000 891.000000
mean      29.6991176  32.204208
sd        14.5264973  49.693429
median    28.0000000  14.454200
trimmed   29.2692308  21.378716
mad       13.3434000  10.236167
min        0.4200000   0.000000
max       80.0000000 512.329200
range     79.5800000 512.329200
skew       0.3874744   4.771210
kurtosis   0.1597671  33.123068
se         0.5436405   1.664792

<p style="font-weight: bold; font-style: italic;">Description of categorical columns in train data</p>


In [19]:
train_df %>% desc(display_stat = "character")


n   n_unique
Survived 891   2     
Pclass   891   3     
Name     891 891     
Sex      891   2     
SibSp    891   7     
Parch    891   7     
Ticket   891 681     
Cabin    204 148     
Embarked 889   4

## 5) **Data split**

<p>Separate target from the data and then split the data into train and validation</p>


In [20]:
# Initial split 🪓
init_split <- train_df %>% initial_split(prop = 3 / 4)


In [29]:
# 75 percent will be for train data
train_data <- init_split %>% training()

# 25 percent will be for test data.
test_data <- init_split %>% testing()

# View the shape of the data.
train_dim <- train_data %>% dim()
test_dim <- test_data %>% dim()


cat(sprintf(
  "Train sample:\n  Rows: %d\n  Columns: %d\n",
  train_dim[1], train_dim[2]
))

cat("\n")

cat(sprintf(
  "Test sample:\n  Rows: %d\n  Columns: %d\n",
  test_dim[1], test_dim[2]
))

Train sample:
  Rows: 668
  Columns: 11

Test sample:
  Rows: 223
  Columns: 11


## 6) **Data cleaning**

<p>Modifying or removing data that is inaccurate, duplicate, incomplete, incorrectly formatted, or corrupted</p>


### i) Check the missing values (NaN)

<p>Search for missing value in the dataframes</p>


In [ ]:
handle_nan <- R6Class(
  "handle_nan",
  private = list(
    train_df = NULL,
    test_df = NULL,
    strategy = NULL,
    fill_value_to_na = function(column_name, value) {
      private$train_df <- private$train_df %>%
        mutate(column_name = replace_na(column_name, value))
    }
  ),
  public = list(
    initialize = function(train_df = NA, test_df = NA, strategy = NA) {
      private$train_df <- train_df
      private$test_df <- test_df
      private$strategy <- strategy
    },
    handle = function() {
      get_names <- names(private$strategy)
      for (i in seq_along(private$strategy)) {
        private$fill_value_to_na(
          get_names[i],
          0
        )
        break
      }
    }
  )
)

fill_data <- handle_nan$new(
  train_data,
  test_data,
  strategy = list(Age = "mean")
)
fill_data$handle()

In [ ]:
train_data %>% fill

In [ ]:
def plot_nan(data: pd.DataFrame, name: str | None = None, only_nan: bool = True, use_percent=False) -> None:
    nan_df = check_n_nan(data, only_nan)
    nan_df['percent_nan'] = nan_df['percent_nan'].str.replace(
        '%', '').astype(float).round().astype(int)

    radius = 5
    y_name = 'percent_nan' if use_percent else 'total_nan'
    startup_title = 'Percentage' if use_percent else 'Number'
    y_title = f'{startup_title} of missing values'
    y_name = alt.Y(y_name).title(y_title)
    name = name if name else 'a dataset'
    title = f'{startup_title} Of Missing values In {name.title()}'
    subtitle = f'A chart represents missing data in {name}'

    base = alt.Chart(
        nan_df,
        title=alt.Title(
            title,
            subtitle=subtitle,
            anchor='start',
        )
    ).mark_bar(
        cornerRadiusTopLeft=radius,
        cornerRadiusTopRight=radius
    ).encode(
        x=alt.X('columns').title('Columns names'),
        y=y_name,
        color=alt.Color('columns').scale(
            scheme='lightgreyred',
            reverse=True
        ).legend(orient='right'),
        tooltip=["columns", "total_nan", 'percent_nan'],
        text='columns'
    ).configure_axis(
        grid=False
    ).configure_view(
        stroke=None
    )

    return base

#### a) Observing missing data in train and validation data


##### One) X train dataset


In [ ]:
check_n_nan(train_x)


In [ ]:
plot_nan(
  train_x,
  name = "x train dataset",
  only_nan = False
)


**conclusion:** X train data sample contains missing data of 21.11%(141) of Age column, 78.29%(523) of cabin column and 0.3%(2) of Embarked column are the amount of missing data in the dataset from the data size of 668.


##### Two) X validation dataset


In [ ]:
check_n_nan(valid_x)


In [ ]:
plot_nan(
  valid_x,
  name = "x validation dataset",
  only_nan = False,
  use_percent = True
)


**conclusion:** In X validation data sample there are 16.14%(36) of Age column and 73.54%(164) of cabin column are the amount of missing data in the dataset from the data size of 223.


#### b) Observing missing data in test dataset


In [ ]:
check_n_nan(test_df)


In [ ]:
plot_nan(
  test_df,
  name = "test dataset",
  only_nan = False,
  use_percent = False
)


**conclusion:** The test data frame contains missing data of 20.57%(86) of Age column, 78.23%(327) of cabin column and 0.24%(1) of Fare column are the amount of missing data in the dataset from the data size of 418.


### ii) Handling the missing values (NaN)

<p>Filling missing value in the dataframes</p>


In [ ]:
class FillNaN:
    def __init__(self, dfs: List[pd.DataFrame], fill_with: Dict[str, Any]) -> None:
        self.dfs = dfs
        self.fill_with = fill_with

    @property
    def fill(self) -> List[pd.DataFrame]:
        """
        Fills missing values in the data frames.
        :return: A tuple of data frames
        """
        lst = []
        for df in self.dfs:
            for col, strat in self.fill_with.items():
                df = self.strategy(df, strat=strat, col_name=col)
            lst.append(df)
        return lst

    @staticmethod
    def strategy(df: pd.DataFrame, strat: str, col_name: str) -> pd.DataFrame:
        match strat:
            case 'mean':
                df[col_name] = df[col_name].fillna(df[col_name].mean())
            case 'mode':
                df[col_name] = df[col_name].fillna(df[col_name].mode()[0])
            case 'min':
                df[col_name] = df[col_name].fillna(df[col_name].min())
            case 'max':
                df[col_name] = df[col_name].fillna(df[col_name].max())
            case 'median':
                df[col_name] = df[col_name].fillna(df[col_name].median())
            case 'bfill':
                df[col_name] = df[col_name].bfill()
            case 'ffill':
                df[col_name] = df[col_name].ffill()
            case "drop":
                df = df.drop(col_name, axis=1)
            case _:
                raise ValueError(
                    f"Ooh, strategy has received invalid value `{strat}` for strat")
        return df

Filling the missing value in age with mean, Fare with ffill, Embarked with mode and Cabin with drop.


In [ ]:
fill = FillNaN(
    [train_x, valid_x, test_df],
    {"Age": "mean",
     "Fare": "ffill",
     "Embarked": "mode",
     "Cabin": "drop"})

In [ ]:
clean_train_x, clean_valid_x, clean_test_df = fill.fill

Check if there still any missing values in the data.


**Clean train x**


In [ ]:
check_n_nan(clean_train_x, only_nan = False)


**Clean valid x**


In [ ]:
check_n_nan(clean_valid_x, only_nan = False)


**Clean test df**


In [ ]:
check_n_nan(clean_test_df, only_nan = False)


### iii) Handling the categorical columns

<p>Encode categorical column to integer and drop high cardinality columns in the dataframes</p>


In [ ]:
class CategoryEncoder:
    def __init__(
            self,
            datasets: List[pd.DataFrame],
            handle_with: Dict[str, str],
    ) -> None:
        self.datasets = datasets
        self.handle_with = handle_with

    def strategy(self, df: pd.DataFrame, col: str, strat: str) -> pd.DataFrame:
        match strat:
            case 'label_encode':
                series = df[col].astype('category').cat.codes
                df[col] = series
            case 'one_hot_encode':
                data = pd.get_dummies(
                    df[col],
                    columns=list(df[col].unique())
                ).astype(int)
                df = pd.concat([df, data], axis=1)
                df = df.drop(col, axis=1)
            case 'drop':
                df = df.drop(col, axis=1)
            case _:
                raise ValueError()

        return df

    def encode(self) -> List[pd.DataFrame]:
        """
        Encode categorical columns to integer.
        :return: A tuple of data frames
        """

        datasets = []

        for dataset in self.datasets:
            for col, strat in self.handle_with.items():
                dataset = self.strategy(dataset, col, strat)
            datasets.append(dataset)

        return datasets

In [ ]:
# First five rows from the clean train x.
clean_train_x.head()


In [ ]:
# Exclude the int and float from the data.
cat_df = clean_train_x.select_dtypes(exclude=[int, float])
cat_df.head()

In [ ]:
# Fetch the categorical columns.
cat_columns <- list(cat_df.columns)
cat_columns


Check for columns with high cardinality column in the data.


In [ ]:
clean_train_x[cat_columns].nunique()

In [ ]:
# Instantiate the CategoryEncoder
encoder = CategoryEncoder(
    [clean_train_x, clean_valid_x, clean_test_df],
    {
        "Sex": "label_encode",
        "Embarked": "one_hot_encode",
        "Name": "drop",
        "Ticket": "drop"
    }
)

# Encode the column
num_train_x, num_valid_x, num_test_df = encoder.encode()

In [ ]:
num_train_x.dtypes


## 7) **Data visualisation**

<p>Represeting the data in graphical way like making chart</p>


In [ ]:
visual_df = pd.concat([clean_train_x, clean_valid_x])
visual_df['Survived'] = np.concatenate([train_y, valid_y])
visual_df.head()

In [ ]:
alt.renderers.enable("html")


class Graph:
    def __init__(
        self, data: pd.DataFrame, map_values: Optional[Dict[str, Dict[str, Any]]] = None
    ) -> None:
        data = data.copy()

        if map_values:
            for key, value in map_values.items():
                data[key] = data[key].map(lambda col: value[col])

        self.__data = data

    def setting_colors(
        self, x: str, color: List[str] | str
    ) -> Tuple[pd.DataFrame, alt.Color]:

        # Color map.
        if isinstance(color, list):
            alt_color = (
                alt.Color(x).scale(range=color, reverse=True).legend(orient="right")
            )
        else:
            alt_color = (
                alt.Color(x).scale(scheme=color, reverse=True).legend(orient="right")
            )
        return alt_color

    def col_bar_char(
        self,
        x_column: str,
        group_by: str,
        color: List | str = "paired",
        width: int = 200,
        height: int = 200,
        title: Optional[str] = None,
        subtitle: Optional[str] = None,
        x_title: str | None = None,
        y_title: str | None = None,
    ):
        alt_color = self.setting_colors(x_column, color=color)

        data = self.__data

        x_title = x_title.title() if x_title else x_column.title()
        y_title = y_title.title() if y_title else "Number of Passengers"

        source = data.groupby(group_by)[x_column].value_counts().reset_index()

        grouped_bars = (
            alt.Chart(source)
            .mark_bar(cornerRadius=10)
            .encode(
                x=alt.X(x_column, title=x_title),
                y=alt.Y("count:Q", title=y_title),
                color=alt_color,
                column=group_by,
            )
            .encode(
                tooltip=[
                    alt.Tooltip(x_column, title=x_title),
                    alt.Tooltip("count", title=y_title),
                ],
            )
            .properties(
                width=width,
                height=height,
                title=alt.Title(
                    title if title else "",
                    fontSize=18,
                    subtitle=subtitle if subtitle else "",
                    subtitleFontSize=12,
                    # dy=-10,
                ),
            )
        )

        return grouped_bars

    def bar_chart(
        self,
        x_column: str,
        color: List | str = "paired",
        width: int = 200,
        height: int = 200,
        title: Optional[str] = None,
        subtitle: Optional[str] = None,
        x_title: str | None = None,
        y_title: str | None = None,
    ) -> alt.ChartDataType:  # Make copy then change column data to str

        alt_color = self.setting_colors(x_column, color=color)

        data = self.__data

        x_title = x_title.title() if x_title else x_column.title()
        y_title = y_title.title() if y_title else "Number of Passengers"

        # Return the alt chart
        bar_layer = (
            alt.Chart(data)
            .mark_bar(cornerRadius=10)
            .encode(
                x=alt.X(
                    x_column,
                    title=x_title,
                    axis=alt.Axis(grid=False, labelFontWeight="bold"),
                ),
                y=alt.Y(
                    "count()",
                    title=y_title,
                    axis=alt.Axis(grid=False, labelFontWeight="bold"),
                ),
                color=alt_color.legend(orient="right"),
                tooltip=[
                    alt.Tooltip(x_column, title=x_title),
                    alt.Tooltip("count()", title=y_title),
                ],
            )
            .properties(
                width=width,
                height=height,
                title=alt.Title(
                    title if title else "",
                    fontSize=18,
                    subtitle=subtitle if subtitle else "",
                    subtitleFontSize=12,
                    # dy=-10,
                ),
            )
        )

        text_layer = (
            alt.Chart(data)
            .mark_text(
                align="center",
                baseline="bottom",  # Align text to the bottom of its bounding box
                dy=-5,  # Adjust vertical position (negative for above the bar)
                fontSize=12,
                fontWeight="bold",
            )
            .encode(x=x_column, y="count()", text=alt.Text("count()"))
        )

        return bar_layer + text_layer  # .configure_view(

    def donut_chart(
        self,
        x: str,
        inner_radius: int = 50,
        color: List | str = "paired",
        width: int = 200,
        height: int = 200,
    ) -> alt.ChartDataType:

        data = self.__data

        value_counts = data[x].value_counts().reset_index()
        value_counts["count"] = (value_counts["count"] / data.shape[0]).round(3)

        donut_chart = (
            alt.Chart(value_counts)
            .mark_arc(innerRadius=inner_radius)
            .encode(
                theta="count",
                color=f"{x}:N",
                tooltip=[
                    alt.Tooltip(f"{x}".title()),
                    alt.Tooltip("count", title="Percentage"),
                ],
            )
            .properties(
                width=width,
                height=height,
            )
        )
        return donut_chart

    def scatter_chart(
        self,
        x: str,
        y: str,
        color_with: Optional[str] = None,
        width: int = 300,
        height: int = 300,
        title: Optional[str] = None,
        subtitle: Optional[str] = None,
        color: List[str] | str = "dark2",
    ):
        alt_color = None

        if color_with:
            alt_color = self.setting_colors(color_with, color=color)

        tooltips = (
            [alt.Tooltip(x), alt.Tooltip(y), alt.Tooltip(color_with)]
            if color_with
            else [alt.Tooltip(x), alt.Tooltip(y)]
        )

        encode_param = (
            dict(x="%s:Q" % x, y="%s:Q" % y, color=alt_color, tooltip=tooltips)
            if color_with
            else dict(x="%s:Q" % x, y="%s:Q" % y, tooltip=tooltips)
        )

        data = self.__data

        # Scatter Plot
        points = (
            alt.Chart(data)
            .mark_point()
            .encode(**encode_param)
            .properties(
                width=width,
                height=height,
                title=alt.Title(
                    title if title else "",
                    fontSize=18,
                    subtitle=subtitle if subtitle else "",
                    subtitleFontSize=12,
                ),
            )
        )

        points = points.configure_view(stroke=None)

        return points

    def concat(
        self,
        *args,
        use: Literal["vconcat", "hconcat"] = "vconcat",
        show_stroke: bool = False,
    ):
        if use == "vconcat":
            if show_stroke:
                return alt.vconcat(*args)
            return alt.vconcat(*args).configure_view(stroke=None)
        elif use == "hconcat":
            if show_stroke:
                return alt.hconcat(*args)
            return alt.hconcat(*args).configure_view(stroke=None)
        else:
            raise ValueError("Received invalid value `%s`" % use)


graph = Graph(
    visual_df,
    map_values={
        "Sex": {0: "Female (0)", 1: "Male (1)"},
        "Survived": {0: "Didn't survive", 1: "survived"},
    },
)
graph.bar_chart("Embarked")
donut = graph.donut_chart("Embarked")
col_bar = graph.scatter_chart(
    "Age", "Fare", color_with="Survived", color=["red", "blue"]
)
col_bar

# hconcat = graph.concat(bar, donut, use="hconcat", show_stroke=True)
# graph.concat(hconcat, col_bar, use="vconcat")
# col_bar

In [ ]:
x = clean_train_x['Embarked'].value_counts().reset_index()
x["count"] = (x["count"] / clean_train_x.shape[0]).round(3)